pkg/apiserver/watch.go

```diff
@@ -64,119 +59,218 @@ func (w *realTimeoutFactory) TimeoutCh() (<-chan time.Time, func() bool) {
 	return t.C, t.Stop
 }
 
+type textEncodable interface {
+	// EncodesAsText should return true if objects should be transmitted as a WebSocket Text
+	// frame (otherwise, they will be sent as a Binary frame).
+	EncodesAsText() bool
+}
+
 // serveWatch handles serving requests to the server
+// TODO: the functionality in this method and in WatchServer.Serve is not cleanly decoupled.
 func serveWatch(watcher watch.Interface, scope RequestScope, req *restful.Request, res *restful.Response, timeout time.Duration) {
-	s, mediaType, err := negotiateOutputSerializer(req.Request, scope.Serializer)
+	// negotiate for the stream serializer
+	serializer, mediaType, err := negotiateOutputSerializer(req.Request, scope.StreamSerializer)
 	if err != nil {
 		scope.err(err, res.ResponseWriter, req.Request)
 		return
 	}
-	// TODO: replace with typed serialization
-	if mediaType != "application/json" {
-		writeRawJSON(http.StatusNotAcceptable, (errNotAcceptable{[]string{"application/json"}}).Status(), res.ResponseWriter)
+	encoder := scope.StreamSerializer.EncoderForVersion(serializer, scope.Kind.GroupVersion())
+
+	useTextFraming := false
+	if encodable, ok := encoder.(textEncodable); ok && encodable.EncodesAsText() {
+		useTextFraming = true
+	}
+
+	// find the embedded serializer matching the media type
+	// TODO: store content type on runtime.Unknown and serialize that into the stream, allowing for mixed payloads
```

Comment:
Yeah but extension does not, so the TODO we left in your PR needs to be
part of this PR I think.

On Mon, Apr 4, 2016 at 12:40 PM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/apiserver/watch.go
> <https://github.com/kubernetes/kubernetes/pull/23806#discussion_r58406427>
> :
>
> >  	if err != nil {
> >  		scope.err(err, res.ResponseWriter, req.Request)
> >  		return
> >  	}
> > -	// TODO: replace with typed serialization
> > -	if mediaType != "application/json" {
> > -		writeRawJSON(http.StatusNotAcceptable, (errNotAcceptable{[]string{"application/json"}}).Status(), res.ResponseWriter)
> > +	encoder := scope.StreamSerializer.EncoderForVersion(serializer, scope.Kind.GroupVersion())
> > +
> > +	useTextFraming := false
> > +	if encodable, ok := encoder.(textEncodable); ok && encodable.EncodesAsText() {
> > +		useTextFraming = true
> > +	}
> > +
> > +	// find the embedded serializer matching the media type
> > +	// TODO: store content type on runtime.Unknown and serialize that into the stream, allowing for mixed payloads
>
> runtime.Unknown already has ContentType, right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23806/files/26e6630171ddc8f3653bbcfa270efb87e1073818#r58406427>
>